### PR TITLE
Enable presets tab to show preset file in explorer

### PR DESCRIPTION
**Pull request by vickyliu-go4it**
With this change, we enabled the presets tab to support showing the
selected file in explorer when user select the "Show in Explorer" action
in the right-click menu.

Test: run unit test
Bug: http://b/187411020

### CODE REVIEW

File: src/OrbitGl/App.cpp [hunk 1]
```diff
@@ -2042,6 +2043,19 @@ void OrbitApp::LoadPreset(const PresetFile& preset_file) {
...
 }
 
+void OrbitApp::ShowPresetInExplorer(const PresetFile& preset) {
+  const auto file_exists = orbit_base::FileExists(preset.file_path());
+  if (file_exists.has_error()) {
+    ERROR("Unable to find preset file: %s", file_exists.error().message());
+    return;
+  }
+
+  std::string command = absl::StrFormat("explorer.exe /select,\"%s\"", preset.file_path().string());
```
Comment (karupayun) [inline src/OrbitGl/App.cpp:2053]: Does this work on Linux?

File: src/OrbitGl/App.cpp [hunk 2]
```diff
@@ -2042,6 +2043,19 @@ void OrbitApp::LoadPreset(const PresetFile& preset_file) {
...
+  if (system(command.c_str()) == 0) return;
```
Comment (karupayun) [inline src/OrbitGl/App.cpp:2054]: Could you add a comment to explain this line?

File: src/DataViews/PresetsDataView.cpp [hunk 3]
```diff
@@ -199,6 +201,13 @@ void PresetsDataView::OnContextMenu(const std::string& action, int menu_index,
                           absl::StrFormat("Could not delete preset \"%s\".", filename));
     }
 
+  } else if (action == kMenuActionShowInExplorer) {
+    if (item_indices.size() != 1) {
```
Comment (karupayun) [inline src/DataViews/PresetsDataView.cpp:205]: So, if `item_indices.size() == 2`, we will show the option but it won't do anything after clicking it?

File: src/DataViews/PresetsDataViewTest.cpp [hunk 4]
```diff
@@ -208,19 +208,20 @@ TEST_F(PresetsDataViewTest, CheckPresenceOfContextMenuEntries) {
 
   // Loadable preset
   EXPECT_THAT(FlattenContextMenuWithGrouping(view_.GetContextMenuWithGrouping(0, {0})),
-              testing::UnorderedElementsAre("Copy Selection", "Export to CSV", "Load Preset",
-                                            "Delete Preset"))
+              testing::ElementsAre("Load Preset", "Delete Preset", "Show in Explorer",
```
Comment (karupayun) [inline src/DataViews/PresetsDataViewTest.cpp:211]: I love this change, but it should belong to another PR where we change all of them in the file.

**Review by karupayun — COMMENTED** [review]
Wild Backlog appeared.
Orbit used Vicky's power, It's super effective.

File: src/OrbitGl/App.cpp (hunk 1 above)
Comment (vickyliu-go4it) [reply src/OrbitGl/App.cpp:2053]: It only works for windows and this is also one of the reason that I add it to `OrbitApp` rather than the file utils in `OrbitBase` which supports both windows and linux.

Also we don't open Orbit client on Linux. So this might be fine XD

File: src/DataViews/PresetsDataView.cpp (hunk 3 above)
Comment (vickyliu-go4it) [reply src/DataViews/PresetsDataView.cpp:205]: The UI enforced this to be a single selection already. So we won't have the case that `item_indices.size() == 2`. We have a comment in `GetContextMenuWithGrouping`. I will also add a comment here.

File: src/DataViews/PresetsDataViewTest.cpp (hunk 4 above)
Comment (vickyliu-go4it) [reply src/DataViews/PresetsDataViewTest.cpp:211]: We can check the order in this PR because we enforce the single selection in the presets tab. However, we support multi-selection in other data views, which means that there can be hundreds of combination cases for the menu content and so we cannot use `ElementsAre` to check the order.

File: src/OrbitGl/App.cpp (hunk 1 above)
Comment (karupayun) [reply src/OrbitGl/App.cpp:2053]: Half the team use Orbit client on Linux. Am I missing something? @beckerhe

File: src/DataViews/PresetsDataView.cpp (hunk 3 above)
Comment (karupayun) [reply src/DataViews/PresetsDataView.cpp:205]: Not sure if I got it 100%, if that was enforced in other part, and we are sure that it couldn't be greater than 1, I would write instead:
`if (item_indices.size() == 0) return;`

But if also it shouldn't be 0 because it always should be a single selection (and that is enforced somewhere else), I would rather CHECK it.
`CHECK (item_indices.size() == 1);`

File: src/OrbitGl/App.cpp (hunk 1 above)
Comment (beckerhe) [reply src/OrbitGl/App.cpp:2053]: Nevertheless we should not deliberately break the Linux UI. We could just make a Windows-Linux-distinction and call a different command on Linux. This one should work:

```
dbus-send --session --print-reply --dest=org.freedesktop.FileManager1 --type=method_call /org/freedesktop/FileManager1 org.freedesktop.FileManager1.ShowItems array:string:"file:///%s" string:""
```

File: src/OrbitGl/App.cpp (hunk 2 above)
Comment (beckerhe) [reply src/OrbitGl/App.cpp:2054]: Using `system` works but is not great from a security perspective as it passes the command string to cmd.exe (to /bin/sh on Linux). There might be ways to inject commands into the preset name which we want to avoid at all costs.

It would be better to use `QProcess::execute` instead. As far as I remember we have a Qt dependency in OrbitGl now anyway, right?

File: src/OrbitGl/App.cpp [hunk 5]
```diff
@@ -2042,6 +2043,19 @@ void OrbitApp::LoadPreset(const PresetFile& preset_file) {
   });
 }
 
+void OrbitApp::ShowPresetInExplorer(const PresetFile& preset) {
+  const auto file_exists = orbit_base::FileExists(preset.file_path());
+  if (file_exists.has_error()) {
+    ERROR("Unable to find preset file: %s", file_exists.error().message());
+    return;
```
Comment (beckerhe) [inline src/OrbitGl/App.cpp:2050]: This should probably be a user-facing error showing a text-box. I think you could use `SendErrorToUi` for that.

File: src/DataViews/PresetsDataViewTest.cpp (hunk 4 above)
Comment (karupayun) [reply src/DataViews/PresetsDataViewTest.cpp:211]: My bad, I saw UnorderedElementsAre in lines 175, 178 and supposed you could have changed it. I still think that this should be in another PR, but I'm not too strong about that.

For other dataviews (discussion for other PR), we can check instead than the current order of actions have the same order than the theoretical order (including all the actions), so if all actions are A - B - C - D - E ...., we don't have something like A - E - C in our array. Let me know if you want to discuss it, a single-for should be enough.

File: src/OrbitGl/App.cpp [hunk 6]
```diff
@@ -2042,6 +2043,19 @@ void OrbitApp::LoadPreset(const PresetFile& preset_file) {
...
+  }
+
+  std::string command = absl::StrFormat("explorer.exe /select,\"%s\"", preset.file_path().string());
+  if (system(command.c_str()) == 0) return;
+
+  ERROR("Unable to show preset file in explorer by running: %s", command);
```
Comment (beckerhe) [inline src/OrbitGl/App.cpp:2056]: I think this should also be surfaced to the user. Maybe we don't wanna show the command in that case, but at least an error message that we were unable to open the file manager.

File: src/OrbitGl/App.cpp (hunk 1 above)
Comment (vickyliu-go4it) [reply src/OrbitGl/App.cpp:2053]: Thanks for the suggestion. If so I will move this method to `OrbitBase/File.h` and make it support both windows and linux

Comment (beckerhe) [reply src/OrbitGl/App.cpp:2053]: I believe we don't have an easy way to start a process in OrbitBase (except using `system` or other low-level system APIs which is not ideal / a lot of work / hard to get right)

Comment (beckerhe) [reply src/OrbitGl/App.cpp:2053]: I'm not opposed to that but you have to willing to spend some time to understand the Posix and Windows APIs to make it work properly. :slightly_smiling_face:

Comment (vickyliu-go4it) [reply src/OrbitGl/App.cpp:2053]: 😂 make sense. I will keep it here

File: src/OrbitGl/App.cpp [hunk 7]
```diff
@@ -2042,6 +2044,36 @@ void OrbitApp::LoadPreset(const PresetFile& preset_file) {
...
+  const QString program{"dbus-send"};
+  const std::string message_content = absl::StrFormat(
+      "/org/freedesktop/FileManager1 org.freedesktop.FileManager1.ShowItems "
+      "array:string:\"file:///%s\" string:\"\"",
+      preset.file_path().string());
+  QStringList arguments;
+  arguments << "--session"
+            << "--print-reply"
+            << "--dest=org.freedesktop.FileManager1"
+            << "--type=method_call" << QString::fromStdString(message_content);
```
Comment (beckerhe) [inline src/OrbitGl/App.cpp:2064]: Nit: I would prefer to use `QStringList::push_back` because it's similar to `std::vector` instead of the streaming operator.

Comment (beckerhe) [reply src/OrbitGl/App.cpp:2064]: Alterantively the braced initialization syntax also works.

File: src/OrbitGl/App.cpp [hunk 8]
```diff
@@ -2042,6 +2044,36 @@ void OrbitApp::LoadPreset(const PresetFile& preset_file) {
...
+    SendErrorToUi("Unable to find preset file: %s", file_exists.error().message());
+    return;
+  }
+
+#if defined(__linux)
+  const QString program{"dbus-send"};
+  const std::string message_content = absl::StrFormat(
+      "/org/freedesktop/FileManager1 org.freedesktop.FileManager1.ShowItems "
+      "array:string:\"file:///%s\" string:\"\"",
+      preset.file_path().string());
```
Comment (beckerhe) [inline src/OrbitGl/App.cpp:2059]: Did you intend to make it one argument? I was expecting this to be split in 4 separate arguments? I could be wrong though.

File: src/OrbitGl/App.cpp [hunk 9]
```diff
@@ -2042,6 +2044,36 @@ void OrbitApp::LoadPreset(const PresetFile& preset_file) {
...
+  QStringList arguments;
+  arguments << "--session"
+            << "--print-reply"
+            << "--dest=org.freedesktop.FileManager1"
+            << "--type=method_call" << QString::fromStdString(message_content);
+#elif defined(_WIN32)
+  const QString program{"explorer.exe"};
+  QStringList arguments;
+  arguments << "/select," << QString::fromStdString(preset.file_path().string());
```
Comment (beckerhe) [inline src/OrbitGl/App.cpp:2068]: Shouldn't this be a single argument.

**Review by beckerhe — APPROVED** [review]
Otherwise LGTM

File: src/OrbitGl/App.cpp (hunk 2 above)
Comment (vickyliu-go4it) [reply src/OrbitGl/App.cpp]: Thanks for the suggestion! Changed to use `QProcess::startDetached` here and added a comment.

File: src/DataViews/PresetsDataView.cpp (hunk 3 above)
Comment (vickyliu-go4it) [reply src/DataViews/PresetsDataView.cpp]: Done! Thanks!

File: src/DataViews/PresetsDataViewTest.cpp (hunk 4 above)
Comment (vickyliu-go4it) [reply src/DataViews/PresetsDataViewTest.cpp:211]: Thanks for the suggestion! 

I have an idea about how to test the action orders. But I would like to do it together with the later refactor of the data views XD

File: src/OrbitGl/App.cpp (hunk 5 above)
Comment (vickyliu-go4it) [reply src/OrbitGl/App.cpp]: Done! Thanks!

File: src/OrbitGl/App.cpp (hunk 6 above)
Comment (vickyliu-go4it) [reply src/OrbitGl/App.cpp]: Done! Thanks!

File: src/OrbitGl/App.cpp (hunk 7 above)
Comment (vickyliu-go4it) [reply src/OrbitGl/App.cpp]: Changed to use braced initialisation! Thanks!

File: src/OrbitGl/App.cpp (hunk 8 above)
Comment (vickyliu-go4it) [reply src/OrbitGl/App.cpp]: I misunderstood the syntax for the `dbus-send` command. Changed now. Thanks!

File: src/OrbitGl/App.cpp (hunk 9 above)
Comment (vickyliu-go4it) [reply src/OrbitGl/App.cpp]: Done. Thanks!

File: src/OrbitGl/App.cpp [hunk 10]
```diff
@@ -2060,8 +2060,8 @@ void OrbitApp::ShowPresetInExplorer(const PresetFile& preset) {
                                  "/org/freedesktop/FileManager1",
                                  "org.freedesktop.FileManager1.ShowItems",
                                  QString::fromStdString(absl::StrFormat(
-                                     "array:string:\"file:///%s\"", preset.file_path().string())),
```
Comment (vickyliu-go4it) [inline src/OrbitGl/App.cpp]: `"array:string:\"file:///%s\""` is fine when I directly type in the command in the terminal. But when running with `QProcess::startDetached`, this format string will result in an additional `"` at the end of the file path.

**Review by karupayun — APPROVED** [review]
LGTM! Thanks for addressing all comments :) :)

File: src/OrbitGl/App.cpp (hunk 10 above)
Comment (beckerhe) [reply src/OrbitGl/App.cpp]: That's expected because the command is not running through a shell. Try this: `"array:string:file:///%s"`

Comment (beckerhe) [reply src/OrbitGl/App.cpp]: Ah. I just saw that you had already changed it. But in your version there are 4 slashes: `file:////` Theoretically 2 should be enough when `%s` is an absolute path.

Comment (vickyliu-go4it) [reply src/OrbitGl/App.cpp]: I tried two slashes, but it does not work... 4 works for me

Comment (beckerhe) [reply src/OrbitGl/App.cpp]: In that case I guess we just leave the 4 slashes. Thank you, Vicky! :-)
